docs/source/schema-directives.md

```diff
@@ -1,179 +1,581 @@
+
+## Implementing schema directives
+
+Since the GraphQL specification does not discuss any specific implementation strategy for directives, it's up to each GraphQL server framework to expose an API for implementing new directives.
+
+If you're using Apollo Server, you are also likely to be using the [`graphql-tools`](https://github.com/apollographql/graphql-tools) npm package, which provides a convenient yet powerful tool for implementing directive syntax: the [`SchemaDirectiveVisitor`](https://github.com/apollographql/graphql-tools/blob/wip-schema-directives/src/schemaVisitor.ts) class.
+
+To implement a schema directive using `SchemaDirectiveVisitor`, simply create a subclass of `SchemaDirectiveVisitor` that overrides one or more of the following visitor methods:
+
+* `visitSchema(schema: GraphQLSchema)`
+* `visitScalar(scalar: GraphQLScalarType)`
+* `visitObject(object: GraphQLObjectType)`
+* `visitFieldDefinition(field: GraphQLField<any, any>)`
+* `visitArgumentDefinition(argument: GraphQLArgument)`
+* `visitInterface(iface: GraphQLInterfaceType)`
+* `visitUnion(union: GraphQLUnionType)`
+* `visitEnum(type: GraphQLEnumType)`
+* `visitEnumValue(value: GraphQLEnumValue)`
+* `visitInputObject(object: GraphQLInputObjectType)`
+* `visitInputFieldDefinition(field: GraphQLInputField)`
+
+By overriding methods like `visitObject`, a subclass of `SchemaDirectiveVisitor` expresses interest in certain schema types such as `GraphQLObjectType` (the first parameter type of `visitObject`).
+
+When `SchemaDirectiveVisitor.visitSchemaDirectives` is called with a `GraphQLSchema` object and a map of visitor subclasses (`{ [directiveName: string]: typeof SchemaDirectiveVisitor }`), visitor methods overridden by those subclasses will be invoked with references to any schema objects that have appropriately named directives attached to them, enabling the visitors to inspect or modify the schema.
+
+For example, here is one possible implementation of the `@deprecated` directive we saw above:
+
+```typescript
+import { SchemaDirectiveVisitor } from "graphql-tools";
+
+class DeprecatedDirective extends SchemaDirectiveVisitor {
+  public visitFieldDefinition(field: GraphQLField<any, any>) {
+    value.isDeprecated = true;
+    value.deprecationReason = this.args.reason;
+  }
+
+  public visitEnumValue(value: GraphQLEnumValue) {
+    value.isDeprecated = true;
+    value.deprecationReason = this.args.reason;
+  }
+}
+```
+
+In order to apply this implementation to a schema that contains `@deprecated` directives, simply pass the `DeprecatedDirective` class to the `makeExecutableSchema` function via the `directives` option:
+
+```typescript
+import { makeExecutableSchema } from "graphql-tools";
+
+const typeDefs = `
+type ExampleType {
+  newField: String
+  oldField: String @deprecated(reason: "Use \`newField\`.")
+}`;
+
+const schema = makeExecutableSchema({
+  typeDefs,
+  directives: {
+    deprecated: DeprecatedDirective
```

Comment:
We should update the API here: https://deploy-preview-640--graphql-tools-docs.netlify.com/docs/graphql-tools/generate-schema.html#makeExecutableSchema

Also we should add an API section at the bottom with an API doc for `SchemaDirectiveVisitor`